src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -872,64 +922,112 @@ CreateDistributedExecution(RowModifyLevel modLevel, List *taskList, bool hasRetu
-		 * the nodes participated in the transaction. And, by re-generating the
-		 * remote task lists during local query execution, we might prevent the adaptive
-		 * executor to kick-in 2PC (or even start coordinated transaction, that's why
-		 * we prefer adding this check here instead of
-		 * Activate2PCIfModifyingTransactionExpandsToNewNode()).
+		 * We prefer to error on any failures for CREATE INDEX
+		 * CONCURRENTLY or VACUUM//VACUUM ANALYZE (e.g., COMMIT_PROTOCOL_BARE).
 		 */
-		if (DistributedExecutionRequiresRollback(execution) || LocalExecutionHappened)
-		{
-			BeginOrContinueCoordinatedTransaction();
+		*errorOnAnyFailure = true;
+		*useTransactionBlocks = TRANSACTION_BLOCKS_DISALLOWED;
+		return;
+	}
 
-			if (TaskListRequires2PC(taskList) || LocalExecutionHappened)
-			{
-				/*
-				 * Although using two phase commit protocol is an independent decision than
-				 * failing on any error, we prefer to couple them. Our motivation is that
-				 * the failures are rare, and we prefer to avoid marking placements invalid
-				 * in case of failures.
-				 */
-				CoordinatedTransactionUse2PC();
+	/*
+	 * In case localExecutionHappened, we force the executor to use 2PC.
+	 * The primary motivation is that at this point we're definitely expanding
+	 * the nodes participated in the transaction. And, by re-generating the
+	 * remote task lists during local query execution, we might prevent the adaptive
+	 * executor to kick-in 2PC (or even start coordinated transaction, that's why
+	 * we prefer adding this check here instead of
+	 * Activate2PCIfModifyingTransactionExpandsToNewNode()).
+	 */
+	if (LocalExecutionHappened)
+	{
+		BeginOrContinueCoordinatedTransaction();
+		CoordinatedTransactionUse2PC();
 
-				execution->errorOnAnyFailure = true;
-			}
-			else if (MultiShardCommitProtocol != COMMIT_PROTOCOL_2PC &&
-					 list_length(taskList) > 1 &&
-					 DistributedExecutionModifiesDatabase(execution))
-			{
-				/*
-				 * Even if we're not using 2PC, we prefer to error out
-				 * on any failures during multi shard modifications/DDLs.
-				 */
-				execution->errorOnAnyFailure = true;
-			}
+		*errorOnAnyFailure = true;
+		*useTransactionBlocks = TRANSACTION_BLOCKS_REQUIRED;
+		return;
+	}
+
+	if (DistributedExecutionRequiresRollback(taskList))
+	{
+		/* make sure we are in a coordinated transaction (we might already be) */
+		BeginOrContinueCoordinatedTransaction();
+
+		if (TaskListRequires2PC(taskList))
+		{
+			/*
+			 * Although using two phase commit protocol is an independent decision than
+			 * failing on any error, we prefer to couple them. Our motivation is that
+			 * the failures are rare, and we prefer to avoid marking placements invalid
+			 * in case of failures.
+			 */
+			CoordinatedTransactionUse2PC();
+
+			*errorOnAnyFailure = true;
+			*useTransactionBlocks = TRANSACTION_BLOCKS_REQUIRED;
+			return;
+		}
+		else if (MultiShardCommitProtocol != COMMIT_PROTOCOL_2PC &&
```

Comment:
For better readability we can have a method like `IsMultiShardModification(modLevel, taskList)`